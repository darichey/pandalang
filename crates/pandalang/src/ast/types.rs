// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
