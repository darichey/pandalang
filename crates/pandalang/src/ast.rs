#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}
