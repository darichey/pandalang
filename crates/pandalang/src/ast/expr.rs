#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
