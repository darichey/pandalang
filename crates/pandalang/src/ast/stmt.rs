use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
