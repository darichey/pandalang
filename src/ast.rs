#[derive(PartialEq, Eq, Debug)]
pub enum Expr {
    Int(i64),
    Var(String),
    BinOp {
        left: Box<Expr>,
        right: Box<Expr>,
        kind: BinOpKind,
    },
    Let {
        name: String,
        value: Box<Expr>,
        body: Box<Expr>,
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}
