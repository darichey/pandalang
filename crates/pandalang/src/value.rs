use crate::ast;

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Value {
    Int(ast::Int),
    Fun(ast::Fun),
}

impl Value {
    pub fn as_expr(self) -> ast::Expr {
        match self {
            Value::Int(int) => ast::Expr::Int(int),
            Value::Fun(fun) => ast::Expr::Fun(fun),
        }
    }
}
