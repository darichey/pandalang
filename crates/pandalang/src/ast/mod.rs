use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
