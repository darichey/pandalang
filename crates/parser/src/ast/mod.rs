use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
