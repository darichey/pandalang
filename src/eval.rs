use std::collections::HashMap;

use crate::ast::{App, BinOp, BinOpKind, Expr, Fun, Int, Var};
use crate::value::Value;

pub struct Env {
    bindings: HashMap<String, Value>,
}

macro_rules! new_env {
    ($($k:expr => $v:expr),* $(,)?) => {{
        Env { bindings: HashMap::from([$(($k.to_string(), Value::Int(Int { n: $v })),)*]) }
    }};
}

impl Env {
    pub fn lookup(&self, x: &String) -> Value {
        self.bindings.get(x).unwrap().clone()
    }
}

pub fn eval(expr: Expr, env: &Env) -> Value {
    match expr {
        Expr::Int(n) => Value::Int(n),
        Expr::Var(Var { name }) => env.lookup(&name),
        Expr::BinOp(BinOp { left, right, kind }) => {
            let f = match kind {
                BinOpKind::Add => std::ops::Add::add,
                BinOpKind::Sub => std::ops::Sub::sub,
                BinOpKind::Mul => std::ops::Mul::mul,
                BinOpKind::Div => std::ops::Div::div,
            };

            let (x, y) = match (eval(*left, env), eval(*right, env)) {
                (Value::Int(Int { n: x }), Value::Int(Int { n: y })) => (x, y),
                _ => panic!("oh god oh fuck"),
            };

            Value::Int(Int { n: f(x, y) })
        }
        Expr::Fun(fun) => Value::Fun(fun),
        Expr::App(App { fun, arg }) => {
            let Fun {
                arg: arg_name,
                body,
            } = match eval(*fun, env) {
                Value::Fun(fun) => fun,
                _ => panic!("oh god oh fuck"),
            };
            let arg = eval(*arg, env).as_expr();

            eval(cas(*body, arg, arg_name), env)
        }
    }
}

fn cas(e1: Expr, e2: Expr, var: String) -> Expr {
    return match e1 {
        Expr::Var(Var { name }) => {
            if var == name {
                e2
            } else {
                Expr::Var(Var { name })
            }
        }
        Expr::Fun(Fun { arg, body }) => {
            if var == arg {
                Expr::Fun(Fun { arg, body })
            } else {
                Expr::Fun(Fun {
                    arg,
                    body: Box::new(cas(*body, e2, var)),
                })
            }
        }
        Expr::App(App { fun, arg }) => Expr::App(App {
            fun: Box::new(cas(*fun, e2.clone(), var.clone())),
            arg: Box::new(cas(*arg, e2, var)),
        }),
        Expr::Int(n) => Expr::Int(n),
        Expr::BinOp(BinOp { left, right, kind }) => Expr::BinOp(BinOp {
            left: Box::new(cas(*left, e2.clone(), var.clone())),
            right: Box::new(cas(*right, e2, var)),
            kind,
        }),
    };
}

#[cfg(test)]
mod tests {
    use super::Env;
    use crate::ast::Int;
    use crate::value::Value;
    use crate::{eval, parser};
    use std::collections::HashMap;

    fn eval(s: String) -> Value {
        eval::eval(
            *parser::parse(s.as_str()).unwrap(),
            &new_env!("x" => 0, "y" => 1, "x'" => 2, "foo" => 3, "a" => 4, "b" => 5, "c" => 6, "d" => 7, "e" => 8),
        )
    }

    #[test]
    fn evals() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            let source = eval(std::fs::read_to_string(&path).unwrap());
            insta::assert_debug_snapshot!(source);
        });
    }
}
