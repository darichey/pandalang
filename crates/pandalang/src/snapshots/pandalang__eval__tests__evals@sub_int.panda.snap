---
source: src/eval.rs
expression: source
input_file: src/snapshot_inputs/bin_op/sub_int.panda
---
Int(
    Int {
        n: 0,
    },
)
